horovod/spark/torch/remote.py

```diff
@@ -454,13 +458,16 @@ def _validate(epoch):
 
 
 def _train_minibatch_fn():
-    def train_minibatch(model, optimizer, transform_outputs, loss_fn, inputs, labels, sample_weights):
-        optimizer.zero_grad()
+    def train_minibatch(model, optimizer, transform_outputs, loss_fn, inputs, labels, sample_weights, backward_passes_per_step, batch_idx):
+        if batch_idx % backward_passes_per_step == 0:
+            if batch_idx != 0:
+                optimizer.step()
+            optimizer.zero_grad()
         outputs = model(*inputs)
         outputs, labels = transform_outputs(outputs, labels)
         loss = loss_fn(outputs, labels, sample_weights)
+        loss.div_(float(backward_passes_per_step))
```

Comment:
Would be great if we do this only for accumulation >1, we dont want to add new computation nodes for default behavior.

Reply:
sometimes we will convert gradients to fp16 for allreduce, this pre-division might have an impact on the numerical precision. I suggest to do this after accumulation i.e before weight update()

Reply:
> sometimes we will convert gradients to fp16 for allreduce, this pre-division might have an impact on the numerical precision. I suggest to do this after accumulation i.e before weight update()

If I understand it correctly, allreduce is applied to gradients only, so I guess we need to do loss div before accumulation in order to obtain the correct gradients?

Reply:
ideally (loss / n) translates to (gradients / n), that's why we can linearly scale learning rate directly when adding more workers, but pre-division might cause underflow because of small gradients. This is minor for now as we dont plan to introduce fp16 allreduce for estimators anytime soon.

Reply:
> ideally (loss / n) translates to (gradients / n), that's why we can linearly scale learning rate directly when adding more workers, but pre-division might cause underflow because of small gradients. This is minor for now as we dont plan to introduce fp16 allreduce for estimators anytime soon.

Thanks for the clarification.